libraries/chain/steem_evaluator.cpp

```diff
@@ -2342,89 +2342,134 @@ namespace golos { namespace chain {
-                        GOLOS_CHECK_LOGIC(op.vesting_shares >= min_delegation,
-                            logic_exception::cannot_delegate_below_minimum,
-                            "Account must delegate a minimum of ${v}",
-                            ("v",min_delegation)("vesting_shares",op.vesting_shares));
-                    });
-                    _db.create<vesting_delegation_object>([&](vesting_delegation_object& o) {
-                        o.delegator = op.delegator;
-                        o.delegatee = op.delegatee;
-                        o.vesting_shares = op.vesting_shares;
-                        o.min_delegation_time = now;
-                    });
-                }
-                _db.modify(delegator, [&](account_object& a) {
-                    a.delegated_vesting_shares += delta;
-                });
-            } else {
-                GOLOS_CHECK_OP_PARAM(op, vesting_shares, {
-                    GOLOS_CHECK_LOGIC(op.vesting_shares.amount == 0 || op.vesting_shares >= min_delegation,
-                        logic_exception::cannot_delegate_below_minimum,
-                        "Delegation must be removed or leave minimum delegation amount of ${v}",
-                        ("v",min_delegation)("vesting_shares",op.vesting_shares));
-                });
-                _db.create<vesting_delegation_expiration_object>([&](vesting_delegation_expiration_object& o) {
-                    o.delegator = op.delegator;
-                    o.vesting_shares = -delta;
-                    o.expiration = std::max(now + STEEMIT_CASHOUT_WINDOW_SECONDS, delegation->min_delegation_time);
-                });
+        void delegate_vesting_shares_with_interest_evaluator::do_apply(const delegate_vesting_shares_with_interest_operation& op) {
+            ASSERT_REQ_HF(STEEMIT_HARDFORK_0_19__756, "delegate_vesting_shares_with_interest_operation");
+
+            const auto& median_props = _db.get_witness_schedule_object().median_props;
+
+            auto add_fields_to_vdo = [&](vesting_delegation_object& o) {
+                o.interest_rate = std::min(op.interest_rate, median_props.max_delegated_vesting_interest_rate);
+            };
+            delegate_vesting_shares(_db, median_props, op, add_fields_to_vdo);
+        }
+
+        void reject_vesting_shares_delegation_evaluator::do_apply(const reject_vesting_shares_delegation_operation& op) {
+            ASSERT_REQ_HF(STEEMIT_HARDFORK_0_19__756, "reject_vesting_shares_delegation_operation");
+
+            const auto& delegator = _db.get_account(op.delegator);
+            const auto& delegatee = _db.get_account(op.delegatee);
+
+            auto delegation = _db.find<vesting_delegation_object, by_delegation>(std::make_tuple(op.delegator, op.delegatee));
+
+            if (delegation == nullptr) {
+                GOLOS_THROW_MISSING_OBJECT("vesting_delegation_object", op.delegator);
             }
 
+            _db.modify(delegator, [&](account_object& a) {
```

Comment:
Reject must freeze delegation before return. This required to avoid case, where delegator asks delegatee to reject delegation to receive it instantly.